src/bootstrap/src/utils/exec.rs

```diff
@@ -230,17 +223,41 @@ pub fn command<S: AsRef<OsStr>>(program: S) -> BootstrapCommand {
 }
 
 /// Represents the output of an executed process.
-#[allow(unused)]
 pub struct CommandOutput {
     status: CommandStatus,
-    stdout: Vec<u8>,
-    stderr: Vec<u8>,
+    stdout: Option<Vec<u8>>,
+    stderr: Option<Vec<u8>>,
 }
 
 impl CommandOutput {
     #[must_use]
-    pub fn did_not_start() -> Self {
-        Self { status: CommandStatus::DidNotStart, stdout: vec![], stderr: vec![] }
+    pub fn did_not_start(stdout: OutputMode, stderr: OutputMode) -> Self {
+        Self {
+            status: CommandStatus::DidNotStart,
+            stdout: match stdout {
+                OutputMode::Print => None,
+                OutputMode::Capture => Some(vec![]),
+            },
+            stderr: match stderr {
+                OutputMode::Print => None,
+                OutputMode::Capture => Some(vec![]),
+            },
+        }
+    }
```

Comment:
If command isn't started, why do we need to pass these arguments? Can't we handle this with some default values?

Reply:
The output mode determines if it is valid to ask for stdout/stderr. E.g. you can execute a command without capturing its stderr, and then try to read the stderr. In that case it's a bug (the code shouldn't just silently return an empty string), and we need to propagate the information whether the output was captured or not so that we're able to find the bug (rather than just returning an empty string).

Reply:
> The output mode determines if it is valid to ask for stdout/stderr.

Function is called `did_not_start`, which makes me think that we shouldn't need to ask for stdout/stderr. Am I missing something here?

Reply:
The command can be used with `.allow_failure()`. So it doesn't matter if it failed to start or failed to execute, if it fails, and the code uses `allow_failure()`, it is able to inspect the result of the command (including stdout/stderr).

We could model this in a different way (e.g. returning `Result` to signify failures of command start/execution), but that would require a large change across bootstrap to handle the potential failures.

Reply:
What would go wrong if we return `Some(vec![])` for both stdout and stderr from `did_not_start` without requiring any argument?

Reply:
Nothing much really, just that if there were some code that does `run_capture_stdout().stderr()`, and it failed to start, the code wouldn't panic (to tell us the bug).